File: CircuitMap_SRN_LO2.m
clc
clear
N=10^4/2;

% theta=0;
g=1;
T=10^9;
M=2000;
time=linspace(0,10^9,10^9+1);
% thetas=[0.5,0.25,0.1,0.075,0.05,0.025,0.01,0.0075,0.005,0.0025,0.001];
thetas=[0.5,0.05,0];

% time=[10^0:10^0:10^1,10^1:10^1:10^2,10^2:10^2:10^3,10^3:10^3:10^4];
for theta=thetas
            x0A=rand(M,N)+1i.*rand(M,N);
            x0B=rand(M,N)+1i.*rand(M,N);

            for i=1:M
                
            x0A(M,:)=x0A(M,:)/sum(abs(x0A(M,:)));
                        x0B(M,:)=x0B(M,:)/sum(abs(x0B(M,:)));
            end
phiA=zeros(M,N)+1i.*zeros(M,N);
phiB=zeros(M,N)+1i.*zeros(M,N);
MeanvalueA=abs(x0A);
%     x0A=zeros(1,N)+1i.*zeros(1,N);
%     x0B=zeros(1,N)+1i.*zeros(1,N);

%     x0A(1)=1+0*1i;

% x0A(1)=rand(1,N)+1i*;
% Meanvalue=zeros(1,5000);
Variance=zeros(1,length(time));
timeind=0;
for t=time

    timeind=timeind+1;
%     for M=1:100
%     Meanvalue(M)=mean(abs(x0A));

%     Variance(timeind)=mean((abs(x0A)-Meanvalue(timeind)).^2);
    
    for n=1:N
        if n==1
            phiA(:,n)=cos(theta)^2*x0A(:,n)-cos(theta)*sin(theta)*x0B(:,N)+sin(theta)^2*x0A(:,n+1)+cos(theta)*sin(theta)*x0B(:,n);
            phiB(:,n)=cos(theta)^2*x0B(:,n)-cos(theta)*sin(theta)*x0A(:,n)+sin(theta)^2*x0B(:,N)+cos(theta)*sin(theta)*x0A(:,n+1);
        elseif n==N
            phiA(:,n)=cos(theta)^2*x0A(:,n)-cos(theta)*sin(theta)*x0B(:,n-1)+sin(theta)^2*x0A(:,1)+cos(theta)*sin(theta)*x0B(:,n);
            phiB(:,n)=cos(theta)^2*x0B(:,n)-cos(theta)*sin(theta)*x0A(:,n)+sin(theta)^2*x0B(:,n-1)+cos(theta)*sin(theta)*x0A(:,1);
            
        else
            phiA(:,n)=cos(theta)^2*x0A(:,n)-cos(theta)*sin(theta)*x0B(:,n-1)+sin(theta)^2*x0A(:,n+1)+cos(theta)*sin(theta)*x0B(:,n);
            phiB(:,n)=cos(theta)^2*x0B(:,n)-cos(theta)*sin(theta)*x0A(:,n)+sin(theta)^2*x0B(:,n-1)+cos(theta)*sin(theta)*x0A(:,n+1);
        end
    end
    x0A=exp(1i.*g.*abs(phiA).^2).*phiA;
    x0B=exp(1i.*g.*abs(phiB).^2).*phiB;
%     end
    
    MeanvalueA=(abs(MeanvalueA)*timeind+abs(x0A))/(timeind+1);
    
    Variance(timeind)=mean((abs(MeanvalueA(:,2))-mean(abs(MeanvalueA(:,2)))).^2);
%     Variance(timeind)=mean((Meanvalue-mean(Meanvalue))).^2;
end
figure(2)
loglog(time,Variance);
save(['/home/pcs/Documents/CircuitMap/data/variance_',num2str(theta),num2str(M),num2str(N),'.mat'],'Variance','-v7.3');
hold on
% legend(num2str(theta,4))
end
% legend([num2str(thetas,4)])
legend('0.5','0.001','0')
% legend('0.5','0.25','0.1','0.075','0.05','0.025','0.01','0.0075','0.005','0.0025','0.001')

File: CircuitMap_SRN_LO2.c
#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#include <complex.h>
#include <time.h>
#include <omp.h>
#define Pi 3.1415926535897932384626433832795
// #define M 10
#define g 1
#define T pow(10, 7)
#define N 5000
#define theta 0
double complex x0A[N], x0B[N];
double complex x0AM[N], x0BM[N];
double complex phiA, phiB;
double Meanvalue[N];
double Variance[10000];
int taxis[10000];
int main()
{
    // printf("%f", T);
    clock_t start_run, finish_run;
    double time_count;
    int n, m, t, Mind, taxisind = 0;
    double tsave = 0, sumMeanvalue, averMeanvalue, sumAvalue, sumBvalue;
    // for (m = 0; m < M; m++)
    // {
    sumAvalue = 0;
    sumBvalue = 0;
    for (n = 0; n < N; n++)
    {
        x0A[n] = rand() + I * rand();
        x0B[n] = rand() + I * rand();
        sumAvalue = sumAvalue + cabs(x0A[n]);
        sumBvalue = sumBvalue + cabs(x0B[n]);
    }
    // }
    // for (m = 0; m < M; m++)
    // {
    for (n = 0; n < N; n++)
    {
        x0A[n] = x0A[n] / sumAvalue;
        x0B[n] = x0B[n] / sumBvalue;
    }
    // printf("%f\n", cabs(x0A[m][1]));
    // }
    // sumAvalue[2] = 0;

    // for (n = 0; n < N; n++)
    // {
    //     sumAvalue[2] = sumAvalue[2] + cabs(x0A[2][n]);
    // }
    // printf("%f", sumAvalue[2]);

    for (n = 0; n < N; n++)
    {
        // x0A[n][0] = 1 + I * 0;
        Meanvalue[n] = cabs(x0A[n]);
    }
    for (t = 0; t < T; t++)
    {
        start_run = clock();
        //#omp parallel for
        // for (Mind = 0; Mind < M; Mind++)
        // {
        // for (n = 0; n < N; n++)
        // {
        //   phiA = 0 + I * 0;
        //   phiB = 0 + I * 0;
        //   x0AM[n] = x0A[Mind][n];
        //   x0BM[n] = x0B[Mind][n];
        // }
        // printf("%f\n", creal(x0A[0][0]));
        for (n = 0; n < N; n++)
        {
            phiA = 0 + I * 0;
            phiB = 0 + I * 0;
            if (n == 0)
            {
                // printf("%f\n", cabs(x0AM[1]));
                phiA = cos(theta) * cos(theta) * x0A[n] - cos(theta) * sin(theta) * x0B[N - 1] + sin(theta) * sin(theta) * x0A[n + 1] + cos(theta) * sin(theta) * x0B[n];
                phiB = cos(theta) * cos(theta) * x0B[n] - cos(theta) * sin(theta) * x0A[n] + sin(theta) * sin(theta) * x0B[N - 1] + cos(theta) * sin(theta) * x0A[n + 1];
            }
            else if (n == N - 1)
            {
                phiA = cos(theta) * cos(theta) * x0A[n] - cos(theta) * sin(theta) * x0B[n - 1] + sin(theta) * sin(theta) * x0A[0] + cos(theta) * sin(theta) * x0B[n];
                phiB = cos(theta) * cos(theta) * x0B[n] - cos(theta) * sin(theta) * x0A[n] + sin(theta) * sin(theta) * x0B[n - 1] + cos(theta) * sin(theta) * x0A[0];
            }
            else
            {
                // printf("%f\n", cabs(x0AM[n]));
                phiA = cos(theta) * cos(theta) * x0A[n] - cos(theta) * sin(theta) * x0B[n - 1] + sin(theta) * sin(theta) * x0A[n + 1] + cos(theta) * sin(theta) * x0B[n];
                phiB = cos(theta) * cos(theta) * x0B[n] - cos(theta) * sin(theta) * x0A[n] + sin(theta) * sin(theta) * x0B[n - 1] + cos(theta) * sin(theta) * x0A[n + 1];
            }
            x0A[n] = exp(I * g * cabs(phiA) * cabs(phiA)) * phiA;
            x0B[n] = exp(I * g * cabs(phiB) * cabs(phiB)) * phiB;
        }
        // for (n = 0; n < N; n++)
        // {
        //   x0A[Mind][n] = exp(I * g * cabs(phiA[n]) * cabs(phiA[n])) * phiA[n];
        //   x0B[Mind][n] = exp(I * g * cabs(phiB[n]) * cabs(phiB[n])) * phiB[n];
        // }
        // }
        for (n = 0; n < N; n++)
        {
            Meanvalue[n] = (cabs(Meanvalue[n]) * t + cabs(x0A[n])) / (t + 1);
        }
        // printf("%f\n", cabs(x0A[1][0]));
        if (t == ceil(1.021 * (tsave + 1)))
        {
            // printf("%d\n", t);
            tsave = 1.021 * (tsave + 1);
            taxis[taxisind] = t;
            sumMeanvalue = 0;
            for (n = 0; n < N; n++)
            {
                sumMeanvalue = sumMeanvalue + cabs(Meanvalue[n]);
            }
            // printf("%f\n", cabs(Meanvalue[1]));
            averMeanvalue = sumMeanvalue / N;
            // printf("%f\n", cabs(averMeanvalue));

            sumMeanvalue = 0;
            for (n = 0; n < N; n++)
            {
                // printf("%f\n", (cabs(Meanvalue[n]) - averMeanvalue) * (cabs(Meanvalue[n]) - averMeanvalue));

                sumMeanvalue = sumMeanvalue + (cabs(Meanvalue[n]) - averMeanvalue) * (cabs(Meanvalue[n]) - averMeanvalue);
            }
            // printf("%f\n", cabs(sumMeanvalue));
            Variance[taxisind] = sumMeanvalue;
            taxisind = taxisind + 1;
        }
        finish_run = clock();
        time_count = (double)(finish_run - start_run) / CLOCKS_PER_SEC;

        printf("%f\n", time_count);

        // printf("\n");
    }

    FILE *fp1;
    fp1 = fopen("HFt0_try2", "w");
    // perror("fopen");

    // FILE *fp2;
    // fp2 = fopen("HFt1", "w");
    // perror("fopen");

    for (n = 0; n < 10000; n++)
    {
        fprintf(fp1, "%d\t", taxis[n]);
        fprintf(fp1, "%.16f\n", Variance[n]);
    }
    fclose(fp1);
    // fclose(fp2);

    // FILE *fp;
    // fp = fopen("samrat.txt", "w");
    // perror("fopen");
    // for (n = 0; n < 10; n++)
    // {
    //   fprintf(fp, "%d\n", 1);
    // }
    // fclose(fp);

    return 0;
}
